server.js

```diff
@@ -220,6 +221,27 @@ app.use(function(req, res, next) {
     next();
 });
 
+// proxy workspaces to remote machines
+const workspaceProxyOptions = {
+    target: 'invalid',
+    ws: true,
+    pathRewrite: {
+        '^/workspace/[0-9]/': '/',
+    },
+    logProvider: _provider => logger,
+    router: async (req) => {
+        let url = 'not-matched';
+        if (/^\/workspace\/0/.test(req.url)) {
+            url = 'http://localhost:8080';
```

Comment:
`localhost` here is just for demo? Proxy/host will be different in practice?

Reply:
Wait, maybe this `localhost` is scoped to `target`.

Reply:
Just for demo. In reality it should be the remote VM that this particular workspace is running on.